### PR TITLE
[Model] Port over CLIPVisionModel for VLMs

### PR DESCRIPTION
**Pull request by ywang96**
This PR picks up the TODO task in #3042 to port over `CLIPVisionModel` from `transformers` for `LlavaForConditionalGeneration` and `LlavaNextForConditionalGeneration`

Note:
- `gelu_quick` CustomOp was added because of `CLIPMLP`.
- We will still use `CLIPAttention` since it doesn't make too much since to use paged attention here, but all other most other layers are done with vLLM primitives.
- This minimal version of `CLIPVisionModel` is only intended to be used within a vision language model, thus the encoder layer forward pass will be done until the `vision_feature_layer`.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/models/clip.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+        self.patch_embedding = nn.Conv2d(
+            in_channels=config.num_channels,
+            out_channels=self.embed_dim,
+            kernel_size=self.patch_size,
+            stride=self.patch_size,
+            bias=False,
+        )
+
+        self.num_patches = (self.image_size // self.patch_size)**2
```
Comment (DarkLight1337) [inline vllm/model_executor/models/clip.py:37]: Can we move this calculation outside the class so that it can be used with my #5214?

File: vllm/model_executor/models/clip.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+    def __init__(self,
+                 config: CLIPVisionConfig,
+                 vision_feature_layer: int = -1,
+                 quant_config: Optional[QuantizationConfig] = None):
+        super().__init__()
+        self.config = config
+        embed_dim = config.hidden_size
+
+        self.embeddings = CLIPVisionEmbeddings(config)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/clip.py:155]: Typo in the attribute name

File: vllm/model_executor/models/clip.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
+"""Minimal implementation of CLIPVisionModel intended to be only used
```
Comment (DarkLight1337) [inline vllm/model_executor/models/clip.py:1]: It would be nice to unit test the model in isolation to ensure its consistency with HF.

File: vllm/model_executor/models/clip.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    """
+    Transformer encoder consisting of `config.num_hidden_layers` self 
+    attention layers. Each layer is a [`CLIPEncoderLayer`].
+
+    Args:
+        config: CLIPConfig
+    """
+
+    def __init__(self,
+                 config: CLIPVisionConfig,
```
Comment (DarkLight1337) [inline vllm/model_executor/models/clip.py:124]: I think we should move the `vision_feature_layer` argument to the `forward` method to better clarify what the return value represents. Ditto for the other `nn.Module`s that use it.

**Review by DarkLight1337 — COMMENTED** [review]
Thanks for the update! I have left some initial comments.

File: vllm/model_executor/models/clip.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                 vision_feature_layer: int = -1,
+                 quant_config: Optional[QuantizationConfig] = None):
+        super().__init__()
+        self.config = config
+        embed_dim = config.hidden_size
+
+        self.embeddings = CLIPVisionEmbeddings(config)
+        self.pre_layrnorm = nn.LayerNorm(embed_dim, eps=config.layer_norm_eps)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/clip.py:156]: Typo in the attribute name

**Review by DarkLight1337 — COMMENTED** [review]
Fixed a comment

File: vllm/model_executor/models/clip.py (hunk 1 above)
Comment (ywang96) [reply vllm/model_executor/models/clip.py:37]: The code for whole `CLIPVisionEmbeddings` class was copied from `transformers`, so I'm not sure if we should change any logic here.

File: vllm/model_executor/models/clip.py (hunk 3 above)
Comment (ywang96) [reply vllm/model_executor/models/clip.py:1]: Sure. The vision language model test should already serve as an end-to-end integration test for `CLIPVisionModel` since it's only intended to be used there, but I can add separate unit tests for this model itself.

File: vllm/model_executor/models/clip.py (hunk 1 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/clip.py:37]: I am thinking of moving the calculation of number of patches to a function in the same module, so I can import it in my PR.

File: vllm/model_executor/models/clip.py (hunk 5 above)
Comment (ywang96) [reply vllm/model_executor/models/clip.py:156]: Funny enough this was the actual name of this parameter (and thus the name of weights as well) - See https://github.com/huggingface/transformers/blob/22b41b3f8a5cdb37e686d18d8d9a24eb98a331ec/src/transformers/models/clip/modeling_clip.py#L823

Comment (DarkLight1337) [reply vllm/model_executor/models/clip.py:156]: I see. Since it would also impact the weight loading, we should not change the name then. Perhaps add a comment in case others also try to fix this error and end up breaking the model.

Comment (ywang96) [reply vllm/model_executor/models/clip.py:159]: Done

File: vllm/model_executor/models/clip.py (hunk 4 above)
Comment (ywang96) [reply vllm/model_executor/models/clip.py:124]: Done

File: vllm/model_executor/models/clip.py (hunk 1 above)
Comment (ywang96) [reply vllm/model_executor/models/clip.py]: Done

**Comment by ywang96** [comment]
@DarkLight1337 After spending some time on this I realized that it's a bit tricky to add unit tests just for `CLIPVisionModel`. This is because setting the vision tower up alone requires some hacks as it uses vLLM primitive layers but we cannot set it up via `LLM`/ `VllmRunner`.

If you have strong opinion about it, I can dig into it bit more. Otherwise I have addressed all other comments!

**Comment by DarkLight1337** [comment]
Since CLIP is being tested indirectly by the other vision models, I guess that's fine for now.

Can you update Phi-3V to also use vLLM's `CLIPVisionModel`?

**Review by DarkLight1337 — APPROVED** [review]
